packages/vue-renderer/src/renderer.js

```diff
@@ -298,64 +298,47 @@ export default class VueRenderer {
     return fn(opts)
   }
 
-  async renderRoute(url, context = {}, retries = 5) {
-    /* istanbul ignore if */
-    if (!this.isReady) {
-      if (this.context.options.dev && retries > 0) {
-        consola.info('Waiting for server resources...')
-        await waitFor(1000)
-        return this.renderRoute(url, context, retries - 1)
-      } else {
-        throw new Error('Server resources are not available!')
-      }
-    }
+  async renderSPA(context) {
+    const content = await this.renderer.spa.render(context)
 
-    // Log rendered url
-    consola.debug(`Rendering url ${url}`)
+    const APP =
+      `<div id="${this.context.globals.id}">${this.context.resources.loadingHTML}</div>` +
+      content.BODY_SCRIPTS
 
-    // Add url and isSever to the context
-    context.url = url
+    // Prepare template params
+    const templateParams = {
+      ...content,
+      APP,
+      ENV: this.context.options.env
+    }
 
-    // Basic response if SSR is disabled or SPA data provided
-    const { req, res } = context
-    const spa = context.spa || (res && res.spa)
-    const ENV = this.context.options.env
-
-    if (!this.SSR || spa) {
-      const {
-        HTML_ATTRS,
-        HEAD_ATTRS,
-        BODY_ATTRS,
-        HEAD,
-        BODY_SCRIPTS,
-        getPreloadFiles
-      } = await this.renderer.spa.render(context)
-      const APP =
-        `<div id="${this.context.globals.id}">${this.context.resources.loadingHTML}</div>` + BODY_SCRIPTS
-
-      const html = this.renderTemplate(false, {
-        HTML_ATTRS,
-        HEAD_ATTRS,
-        BODY_ATTRS,
-        HEAD,
-        APP,
-        ENV
-      })
+    // Call spa:templateParams hook
+    this.context.nuxt.callHook('vue-renderer:spa:templateParams', templateParams)
+
+    // Render with SPA template
+    const html = this.renderTemplate(false, templateParams)
 
-      return { html, getPreloadFiles: this.getPreloadFiles.bind(this, { getPreloadFiles }) }
+    return {
+      html,
+      getPreloadFiles: this.getPreloadFiles.bind(this)
     }
+  }
 
-    let APP
+  async renderSSR(context) {
     // Call renderToString from the bundleRenderer and generate the HTML (will update the context as well)
-    if (req && req.modernMode) {
-      APP = await this.renderer.modern.renderToString(context)
-    } else {
-      APP = await this.renderer.ssr.renderToString(context)
-    }
+    const renderer = context.modren ? this.renderer.modern : this.renderer.ssr
+    let APP = await renderer.renderToString(context)
+
+    // Call ssr:context hook
+    this.context.nuxt.callHook('vue-renderer:ssr:context', context)
+    await this.context.nuxt.callHook('render:routeContext', context.nuxt) // Legacy
```

Comment:
Why not having a mapping inside `callHook`that when `vue-renderer:ssr:context` is called, also call `render:routeContext`? Just exploring some idea to move legacy code somewhere else?

Reply:
Indeed we have this mapping:

https://github.com/nuxt/nuxt.js/blob/324135df5128957041131e96deb6b8122a88f4c7/packages/core/src/nuxt.js#L28

But params are different.